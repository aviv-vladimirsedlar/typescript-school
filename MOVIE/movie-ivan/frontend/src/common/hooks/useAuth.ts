import { useSelector } from 'react-redux';

import { RootState } from '../../config/store';

export const useAuth = () => {
  return useSelector((state: RootState) => state.auth.isAuthenticated);
};
