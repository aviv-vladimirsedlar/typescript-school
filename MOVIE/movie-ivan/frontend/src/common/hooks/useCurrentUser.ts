import { useState } from 'react';
import { useQuery } from 'react-query';
import { useDispatch } from 'react-redux';

import axiosInstance from '../../config/api';
import { loginSuccess } from '../../config/slices/auth.slice';

const getCurrentUser = async () => {
  const response = await axiosInstance.get('/users/me');
  return response.data;
};

export const useCurrentUser = () => {
  const [currentUser, setCurrentUser] = useState(null);
  const dispatch = useDispatch();

  const { isLoading } = useQuery('currentUser', getCurrentUser, {
    retry: false,
    onSuccess: (res) => {
      setCurrentUser(res);
      dispatch(loginSuccess({ user: res }));
    },
    onError: (error) => {
      console.error('Login error:', error);
    },
  });

  return { currentUser, isLoading };
};
