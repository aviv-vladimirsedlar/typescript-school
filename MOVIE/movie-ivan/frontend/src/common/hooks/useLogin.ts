import { useMutation } from 'react-query';
import { useDispatch } from 'react-redux';
import { useNavigate } from 'react-router-dom';

import axiosInstance from '../../config/api';
import { loginSuccess } from '../../config/slices/auth.slice';

interface LoginData {
  email: string;
  password: string;
}

const loginRequest = async (data: LoginData) => {
  const response = await axiosInstance.post('/users/login', data);
  return response.data;
};

export const useLogin = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  return useMutation(loginRequest, {
    onSuccess: (data) => {
      dispatch(loginSuccess({ user: data.user }));
      navigate('/');
    },
  });
};
