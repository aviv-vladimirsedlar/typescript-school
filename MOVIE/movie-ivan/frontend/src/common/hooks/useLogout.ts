import { useMutation } from 'react-query';
import { useDispatch } from 'react-redux';
import { useNavigate } from 'react-router-dom';

import axiosInstance from '../../config/api';
import { logoutSuccess } from '../../config/slices/auth.slice';

const logoutRequest = async () => {
  const response = await axiosInstance.delete('/users/logout');
  return response.data;
};

export const useLogout = () => {
  const navigate = useNavigate();
  const dispatch = useDispatch();

  return useMutation(logoutRequest, {
    onSuccess: () => {
      dispatch(logoutSuccess());
      setTimeout(() => {
        navigate('/auth/login');
      }, 1000);
    },
    onError: (error) => {
      console.error('Login error:', error);
    },
  });
};
