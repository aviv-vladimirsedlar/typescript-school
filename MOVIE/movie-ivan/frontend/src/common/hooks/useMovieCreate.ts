import { useMutation } from 'react-query';

import axiosInstance from '../../config/api';
import { Movie } from '../types/movie.types';

type MovieData = Pick<Movie, 'title' | 'description' | 'year' | 'duration'>;

const createMovie = async (data: { movieId: string; data: MovieData }) => {
  const response = await axiosInstance.post(`/movies/create`, data.data);
  return response.data;
};

export const useMovieCreate = () => {
  return useMutation(createMovie, {
    onError: (error) => {
      console.error('Movie create error:', error);
    },
  });
};
