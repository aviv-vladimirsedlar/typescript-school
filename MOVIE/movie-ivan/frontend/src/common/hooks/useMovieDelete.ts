import { useMutation } from 'react-query';

import axiosInstance from '../../config/api';

const deleteMovie = async (movieId: string) => {
  const response = await axiosInstance.delete(`/movies/${movieId}`);
  return response.data;
};

export const useMovieDelete = ({ onSuccess }: { onSuccess: () => void }) => {
  return useMutation(deleteMovie, {
    onSuccess,
    onError: (error) => {
      console.error('Delete error:', error);
    },
  });
};
