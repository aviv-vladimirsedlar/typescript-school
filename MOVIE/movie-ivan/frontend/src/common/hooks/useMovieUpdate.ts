import { useMutation } from 'react-query';

import axiosInstance from '../../config/api';
import { Movie } from '../types/movie.types';

type MovieData = Pick<Movie, 'title' | 'description' | 'year' | 'duration'>;

const updateMovie = async (data: { movieId: string; data: MovieData }) => {
  const response = await axiosInstance.put(`/movies/${data.movieId}`, data.data);
  return response.data;
};

export const useMovieUpdate = () => {
  return useMutation(updateMovie, {
    onError: (error) => {
      console.error('Movie update error:', error);
    },
  });
};
