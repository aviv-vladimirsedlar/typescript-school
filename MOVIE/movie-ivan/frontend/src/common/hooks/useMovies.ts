import { useQuery } from 'react-query';
import { useDispatch, useSelector } from 'react-redux';

import axiosInstance from '../../config/api';
import { updateMovieList } from '../../config/slices/movie.slice';
import { RootState } from '../../config/store';

const getMovies = async () => {
  const response = await axiosInstance.get('/movies?limit=100');
  return response.data;
};

export const useMovies = () => {
  const moviesData = useSelector((state: RootState) => {
    return state.movie;
  });

  const dispatch = useDispatch();

  const { isLoading, refetch } = useQuery('getMovies?limit=100', getMovies, {
    retry: false,
    onSuccess: (data) => {
      dispatch(updateMovieList(data));
    },
    onError: (error) => {
      console.error('Movie list error:', error);
    },
  });

  return { data: moviesData.data, meta: moviesData.meta, isLoading, refetch };
};
