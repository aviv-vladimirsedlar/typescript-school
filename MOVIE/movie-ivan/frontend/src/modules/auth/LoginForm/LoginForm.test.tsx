import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import React from 'react';

import '@testing-library/jest-dom';
import { useLogin } from '../../../common/hooks/useLogin';

import { LoginForm } from '.';

jest.mock('../../../common/hooks/useLogin');

const mockMutate = jest.fn();
const mockUseLogin = useLogin as jest.MockedFunction<typeof useLogin>;

describe('LoginForm', () => {
  beforeEach(() => {
    // Reset mocks before each test
    jest.clearAllMocks();
    mockUseLogin.mockReturnValue({
      mutate: mockMutate,
      isLoading: false,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any);
  });

  it('should match snapshot when initially rendered', () => {
    const { asFragment } = render(<LoginForm />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the login form', () => {
    mockMutate.mockImplementation((data, { onSuccess }) => onSuccess && onSuccess());
    render(<LoginForm />);
    expect(screen.getByTestId('email-input')).toBeInTheDocument();
    expect(screen.getByTestId('password-input')).toBeInTheDocument();
    expect(screen.getByTestId('btn-login')).toBeInTheDocument();
  });

  it('shows an error message on invalid credentials', async () => {
    mockMutate.mockImplementation(
      (data, { onError }) => onError && onError({ response: { data: { message: 'Invalid credentials' } } }),
    );
    render(<LoginForm />);

    fireEvent.change(screen.getByTestId('email-input'), { target: { value: 'wrong@aviv-group.com' } });
    fireEvent.change(screen.getByTestId('password-input'), { target: { value: 'wrongpassword' } });
    fireEvent.click(screen.getByTestId('btn-login'));

    await waitFor(() => {
      expect(screen.getByTestId('login-error-message')).toHaveTextContent('Invalid credentials');
    });
  });

  it('does not show an error message on successful login', async () => {
    mockMutate.mockImplementation((data, { onSuccess }) => onSuccess && onSuccess());
    render(<LoginForm />);

    fireEvent.change(screen.getByTestId('email-input'), { target: { value: 'admin@aviv-group.com' } });
    fireEvent.change(screen.getByTestId('password-input'), { target: { value: 'Test@#12345' } });

    fireEvent.click(screen.getByTestId('btn-login'));

    await waitFor(() => {
      expect(screen.queryByTestId('login-error-message')).toBeNull();
    });
  });
});
