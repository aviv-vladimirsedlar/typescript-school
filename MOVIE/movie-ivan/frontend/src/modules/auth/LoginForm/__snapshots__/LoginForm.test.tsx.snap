// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should match snapshot when initially rendered 1`] = `
<DocumentFragment>
  <form
    class="mx-auto w-full max-w-2xl"
  >
    <div
      class="css-167vhyo"
    >
      <div
        class="css-ow2jnx"
        data-testid="email-input"
      >
        <div
          class="css-j7qwjs"
        >
          <div
            class="css-6xq2vb"
          >
            <label
              class="css-1azm4t"
              for="react-aria-:r0:"
              id="react-aria-:r1:"
            >
              Your email
              <span
                class="css-uv7gtw"
              >
                *
              </span>
            </label>
          </div>
        </div>
        <div
          class="css-1c1vv48"
        >
          <input
            aria-labelledby="react-aria-:r1:"
            aria-required="true"
            class="css-pcexyr"
            id="react-aria-:r0:"
            inputmode="text"
            size="1"
            type="text"
            value="admin@aviv-group.com"
          />
        </div>
      </div>
    </div>
    <div
      class="css-1gwlgto"
    >
      <div
        class="css-ow2jnx"
        data-testid="password-input"
      >
        <div
          class="css-j7qwjs"
        >
          <div
            class="css-6xq2vb"
          >
            <label
              class="css-1azm4t"
              for="react-aria-:r5:"
              id="react-aria-:r6:"
            >
              Your password
              <span
                class="css-uv7gtw"
              >
                *
              </span>
            </label>
          </div>
        </div>
        <div
          class="css-1c1vv48"
        >
          <input
            aria-labelledby="react-aria-:r6:"
            aria-required="true"
            class="css-pcexyr"
            id="react-aria-:r5:"
            inputmode="none"
            name="password"
            size="1"
            type="text"
            value="Test@#12345"
          />
        </div>
      </div>
    </div>
    <button
      class="css-1bzl4wl"
      data-testid="btn-login"
      type="button"
    >
      <div
        class="css-o395q5"
      >
        <span
          class="css-1gur7lg"
        >
          Submit
        </span>
      </div>
    </button>
  </form>
</DocumentFragment>
`;
